### PR TITLE
Add cifar10 example using TE+DeepSpeed-Zero+MS-AMP

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Add cifar10 example using TE+DeepSpeed-Zero+MS-AMP
There is a 1P customer who only uses DeepSpeed ZeRO and the model is transformer based. Purely replacing Linear with FP8Linear will not get performance gain. So we suggest them to use TransformerEngine for speeding up and MS-AMP for saving memory. This is a good reference to them.

**Major Revision**
- Add example
- Add document
- Fix a bug in mnist_ddp example

### CODE REVIEW

File: examples/cifar10_deepspeed_te.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""The deepspeed cifar10 exampe using MS-AMP. It is adapted from official deepspeed example.
+
+The only change is add "from msamp import deepspeed" and use FP8VisionTransformer instead of orignial model Net.
```
Comment (wkcn) [inline examples/cifar10_deepspeed_te.py:6]: Please explain the source of model definition and training script.
For example, `Adapted from xxx`

File: examples/cifar10_deepspeed_te.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+            layer_type='encoder',
+            init_method=init_method,
+            output_layer_init_method=init_method,
+            drop_path_rate=drop_path,
+            fuse_qkv_params=True,
+        )
+
+    def forward(self, x):
+        """Forward computation."""
+        # x: seq_len, batch_size, dim
```
Comment (wkcn) [inline examples/cifar10_deepspeed_te.py:151]: The code L151-159 can be simplified with `return self.m(x, attention_mask=None)`

File: msamp/deepspeed/runtime/config.py [hunk 3]
```diff
@@ -12,6 +12,8 @@
 MSAMP_ENABLED_DEFAULT = False
 MSAMP_OPTLEVEL = 'opt_level'
 MSAMP_OPTLEVEL_DEFAULT = 'O1'
+MSAMP_USETE = 'use_te'
```
Comment (wkcn) [inline msamp/deepspeed/runtime/config.py:15]: Please rename it with `MSAMP_USE_TE` and `MSAMP_USE_TE_DEFAULT`.

File: msamp/deepspeed/runtime/config.py [hunk 4]
```diff
@@ -26,6 +28,7 @@ def _initialize_params(self, param_dict):
         super()._initialize_params(param_dict)
         self.msamp_enabled = get_msamp_enabled(param_dict)
         self.msamp_optlevel = get_msamp_optlevel(param_dict)
+        self.msamp_usete = get_msamp_usete(param_dict)
```
Comment (wkcn) [inline msamp/deepspeed/runtime/config.py:31]: `self.msamp_use_te` or `self.msamp_te_enabled`

File: msamp/deepspeed/runtime/config.py (hunk 3 above)
Comment (wkcn) [reply msamp/deepspeed/runtime/config.py:15]: Other choices: `MSAMP_TE_ENABLED`

File: examples/cifar10_deepspeed_te.py (hunk 2 above)
Comment (tocean) [reply examples/cifar10_deepspeed_te.py:151]: Without padding, it will crash.
image

File: examples/cifar10_deepspeed_te.py (hunk 1 above)
Comment (tocean) [reply examples/cifar10_deepspeed_te.py]: Done. Explained the model definition in comment.

File: msamp/deepspeed/runtime/config.py (hunk 3 above)
Comment (tocean) [reply msamp/deepspeed/runtime/config.py]: Replaced `MSAMP_USETE` and `MSAMP_USETE_DEFAULT` with `MSAMP_USE_TE` and `MSAMP_USE_TE_DEFAULT` respectively.

File: msamp/deepspeed/runtime/config.py (hunk 4 above)
Comment (tocean) [reply msamp/deepspeed/runtime/config.py]: Changed from `self.msamp_usete` to `self.msamp_use_te`

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

**Review by guoshzhao — APPROVED** [review]
